### ISSUE TITLE
pip install annoy not working on macOS mojave

### ISSUE THREAD

**Issue by thebarbershop** [spotify/annoy#349]
This is where `pip install annoy` fails:

```bash
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/taegyung/miniconda3/include -arch x86_64 -I/Users/taegyung/miniconda3/include -arch x86_64 -I/Users/taegyung/miniconda3/include/python3.6m -c src/annoymodule.cc -o build/temp.macosx-10.7-x86_64-3.6/src/annoymodule.o -O3 -ffast-math -fno-associative-math -march=native
  warning: include path for stdlibc++ headers not found; pass '-std=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
  In file included from src/annoymodule.cc:15:
  src/annoylib.h:49:10: fatal error: 'vector' file not found
  #include <vector>
           ^~~~~~~~
  1 warning and 1 error generated.
  error: command 'gcc' failed with exit status 1
```

I tried some things I found from previous issues and some general solutions accros the web, including

- sudo pip install annoy
- Installing XCode 9.4 (I wasn't using any version of XCode previously)
- xcode-select --install
- brew remove gcc; brew install gcc

I finally got it by running `conda install -c akode annoy`, but I'm not sure if this is the right way to do it (and don't know why this even works.)

**Comment by erikbern** [comment]
that's wacky. not sure what's going on

**Comment by thebarbershop** [comment]
I couldn't reproduce the situation on another mac (becauase I have only one), but #334 seems like a similar issue. It was resolved by removing the include in problem, but is that going to work here?

**Comment by a1k0n** [comment]
The warning
```
warning: include path for stdlibc++ headers not found
```

hints at the root of the problem; `vector` is a basic C++ standard library header; for some reason, whatever compiler pip install is using doesn't have a C++ compiler properly set up. Hard to say; is there homebrew on that system? which python/pip is it using? which compiler? is gcc an alias for clang or is gcc actually installed somehow?

Anyway it's definitely not something annoy can fix.

**Comment by erikbern** [comment]
well... i could fix it by providing a precompiled wheel i guess. maybe worth doing at some point

**Comment by thebarbershop** [comment]
@a1k0n 
> Is there homebrew on that system? which python/pip is it using? which compiler? is gcc an alias for clang or is gcc actually installed somehow?

I should have gave all these info already. Python is 3.6.8, pip is 18.1.

```bash
> $ /usr/bin/gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/c++/4.2.1
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin18.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

(`> gcc --version` gives the same, as `whereis gcc` throws `/usr/bin/gcc`.)

So `gcc` runs clang, not the gcc proper. Re-linking this doesn't seem to be a good idea, as I think that shouldn't be the process for all mac users using pip to install annoy.

`brew install gcc; brew info gcc` throws:

```
gcc: stable 8.2.0 (bottled), HEAD
GNU compiler collection
https://gcc.gnu.org/
/usr/local/Cellar/gcc/8.2.0 (1,413 files, 295.9MB) *
....(and some more lines)....
```

So gcc proper is installed over there, namely at `/usr/local/Cellar/gcc/8.2.0/bin/gcc-8` and `/usr/local/Cellar/gcc/8.2.0/bin/g++-8`, but I don't think pip isn't using it. The `--version` for each says, respectively,
```
gcc-8 (Homebrew GCC 8.2.0) 8.2.0
g++-8 (Homebrew GCC 8.2.0) 8.2.0
```

[GNU](https://gcc.gnu.org/install/binaries.html) directs HomeBrew for download on mac, so I understand these are gcc proper.

I wrote a simple code snippet, `#include<vector> int main() { return 0; }` and saved it as `test.cpp` and `test.c`. All of `(gcc, g++, /usr/local/Cellar/gcc/8.2.0/bin/gcc-8, /usr/local/Cellar/gcc/8.2.0/bin/g++-8) test.cpp` compiles, and all of `(gcc, g++, /usr/local/Cellar/gcc/8.2.0/bin/gcc-8, /usr/local/Cellar/gcc/8.2.0/bin/g++-8) test.c` fails, so this is not an issue that `vector` is missing in my compiler.

I think it is an issue related to the [deprecation of `libstdc++` by Apple](https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes?language=objc) (see Deprecation Notes), and will affect all future mac users.

**Comment by thebarbershop** [comment]
And I finally got it.

```
env CXX=/usr/local/Cellar/gcc/8.2.0/bin/g++-8 CC=/usr/local/Cellar/gcc/8.2.0/bin/gcc-8 pip install annoy
```
does the job to force pip to use the brew-installed compiler for command line instructions `g++` and `gcc`.

Maybe you don't have to re-write or pre-compile things but just let mac users (after XCode 10 and their clang deprecation of `libstdc++`) know from README.md that they should do this to install annoy.

```
> brew install gcc
> brew info gcc
(and get the path to the brew-installed gcc bin)
> env CXX=/your/brew-installed/gcc/version-number/bin/g++-8\ 
CC=/your/brew-installed/gcc/version-number/bin/gcc-8\
 pip install annoy
```

**Comment by erikbern** [comment]
it's possible i can add something to setup.py to fix this. let me see

**Comment by KatyaRuth** [comment]
Still having this issue. Have tried to set env variables, use conda install, etc., no luck. Having in setup.py would be much appreciated!

**Comment by erikbern** [comment]
@KatyaRuth thanks for reminding me, let me see if i can fix

**Comment by erikbern** [comment]
Something is broken with installing Mojave on my laptop, so I can't test it myself, but I'll try to just fix it and maybe one of you could check that it's working

**Comment by erikbern** [comment]
@thebarbershop @KatyaRuth can you take a look at #362 to see if it fixes the issue?

I'll probably just merge it later today, seems highly likely that it will resolve the issue

**Comment by thebarbershop** [comment]
Here's what I did to check this. Never done this before, so I want to make sure if I did it right.

0. All these being done in a new conda environment, 
1. Brew-uninstalled gcc, checked both commands `gcc` and `g++` run clang.
2. `git clone -b -b fix-349 git@......annoy.git`, then `pip install ./annoy`
3. Ran some python scripts of mine that depends on Annoy.
4. Works great.

**Comment by erikbern** [comment]
This has been fixed in 1.15.1 and published to pypi: https://pypi.org/project/annoy/

**Comment by pfdamasceno** [comment]
For the record, this can also be caused by the need to reinstall `xcode developer tools` after an upgrade. I fixed this problem by running `xcode-select --install` as suggested [here](https://apple.stackexchange.com/a/254381)
